markov-analysis.py

```diff
@@ -0,0 +1,220 @@
+    text_length: int, the length of the random text
+    title: string, title of the source book
+
+    int, int -> str
+    '''
+    word_list = process_file(title, skip_header=True)
+    prefixes = prefix_map(word_list, prefix_length)
+    return list_to_string(random_text(prefixes, prefix_length, text_length))
+
+def list_to_string(word_list):
+    '''
+    Converts a list of words to a string, with spaces between each word and a 
+    period at the end.
+
+    word_list: list
+
+    list -> str
+    '''
+    s = ''
+    for word in word_list:
+        s = s + ('%s ' % word)
+    return (s.strip() + '.')
+
+def random_text(prefix_map, prefix_length, text_length):
+    '''
+    Returns a list of words randomly selected by prefix using the given 
+    prefix map.
+
+    # TODO: This looks messy, is there a better way than changing so many types?
+
+    prefix_map: dict of tuples to lists
+    prefix_length: int, the length (in words) of the sequence of prefixes to use
+    m: int, the length (in words) of the list of random words
+
+    dict, int, int -> list
+    '''
+    random_text = list(random.choice(list(prefix_map.keys())))
+
+    while (len(random_text) < text_length):
+        prefix = tuple(random_text[(len(random_text)-prefix_length):len(random_text)])
+        random_text.append(random.choice(prefix_map[prefix]))
+
+    return random_text
+
+def prefix_map(word_list, n):
+    '''
+    Takes a list of words and returns a dict mapping sequences of n words (which
+    we call the prefixes) to a list of all next words (which we call the 
+    suffixes).
+
+    word_list: list
+    n: int, the number of prefixes to consider
+    
+    list, int -> dict
+    '''
+    ''' TODO: This doesn't do quite what I want but can't figure out why
+        prefix_map = {(''):{word_list[0]}} # First word follows nothing
+        if n > 1: # Map starting prefix sequences of length <n
+            for i in range(1, n):
+                prefix_seq = tuple(word_list[1:i])
+                prefix_map[prefix_seq] = {word_list[i]}
+    '''
+    prefix_map = {}
+    # Maps prefix n-tuples to suffix list
+    for prefix_start in range(0, len(word_list)-n):
+        prefix_seq = tuple(word_list[prefix_start:prefix_start+n]) 
+        suffix = word_list[prefix_start+n]
+        prefix_map[prefix_seq] = prefix_map.get(prefix_seq, []) + [suffix]
+    
+    return prefix_map
+
+# WARNING: skip_header=True only known to work with emma.txt
+def process_file(filename, skip_header):
+    """
+    (Taken from chapter 13 of ThinkPython, modified to return list)
+    Makes a list that contains the words from a file.
+
+    filename: string
+    skip_header: boolean, whether to skip the Gutenberg header
+   
+    returns: list of all words in the file in order of appearance
+    """
+    word_list = []
+    fp = open(filename)
+
+    if skip_header:
+        skip_gutenberg_header(fp)
+
+    for line in fp:
+        process_line(line, word_list)
```

Comment:
my personal preference here would be to rewrite in a style more like:

`word_list += process_line(line)` 

or something.

implicit mutation of arguments is sketchy.

Reply:
done. i've rewritten `process_file` and `process_line` to avoid mutation, and changed their names to `file_to_list` and `line_to_list` respectively